### PR TITLE
feat: add sse & multipart transports to federated dev

### PR DESCRIPTION
**Pull request by obmarg**
Adds [Incremental Delivery][1] & [GraphQL Over SSE][2] support to federated dev.

We already had code for these in the CLI, so I'm mostly just hooking things up.

Fixes GB-5797

[1]: https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md
[2]: https://github.com/graphql/graphql-over-http/blob/main/rfcs/GraphQLOverSSE.md

### CODE REVIEW

File: engine/crates/async-runtime/src/stream.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    stream::{self, BoxStream},
+    Future, FutureExt, Stream, StreamExt,
+};
+
+/// Combines a stream and a "producer" future into a single stream
+///
+/// This can be used when you have the receivng side of a channel and a future that sends
+/// on that channel - combining the two into a single stream that'll run till the channel
+/// is exhausted.  If you drop the stream you also cancel the underlying process.
+pub fn producer_stream<'a, S, P, Item>(stream: S, producer: P) -> impl Stream<Item = Item> + 'a
```
Comment (obmarg) [inline engine/crates/async-runtime/src/stream.rs:12]: Any name suggestions for this welcome, it's an awkward concept to describe succinctly

**Review by Finistere — APPROVED** [review]

File: engine/crates/async-runtime/src/stream.rs (hunk 1 above)
Comment (Finistere) [reply engine/crates/async-runtime/src/stream.rs:12]: I think I would create an extension like:
```
impl Stream {
  fn with_producer_future(&self, producer: P)
}
```

Comment (Finistere) [reply engine/crates/async-runtime/src/stream.rs:12]: Or maybe `join_with_future`

Comment (Finistere) [reply engine/crates/async-runtime/src/stream.rs:12]: "join" feels a bit more idiomatic imho

**Pull request merged**